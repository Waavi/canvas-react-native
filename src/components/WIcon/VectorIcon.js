import { createIconSetFromIcoMoon } from 'react-native-vector-icons'
import icoMoonConfig from './config.json'

// If you need add more Icons use this tutorial - https://medium.com/bam-tech/add-custom-icons-to-your-react-native-application-f039c244386c
export const VectorIcon = createIconSetFromIcoMoon(icoMoonConfig)
