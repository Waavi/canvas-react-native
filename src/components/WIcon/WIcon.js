import React from 'react'
import { View, StyleSheet } from 'react-native'
import PropTypes from '#propTypes'
import { Metrics, Colors } from '@/theme'
import { VectorIcon } from './VectorIcon'

WIcon.propTypes = {
	size: PropTypes.number,
	width: PropTypes.number,
	color: PropTypes.string,
	style: PropTypes.number,
}
WIcon.defaultProps = {
	size: 20,
	width: undefined,
	color: undefined,
	style: undefined,
}

export function WIcon({ size = 20, width = undefined, color = undefined, style, ...iconProps }) {
	const iconSize = Metrics.scale.horizontal(size)
	const iconColor = color || (style && StyleSheet.flatten(style).color) || Colors.black
	if (width) {
		const iconWidth = width ? Metrics.scale.horizontal(width) : iconSize
		const iconMarginLeft = iconWidth !== iconSize ? (iconWidth - iconSize) / 2 : 0
		const defaultStyle = { width: iconWidth, overflow: 'hidden' }
		return (
			<View style={[defaultStyle, style]}>
				<VectorIcon
					size={iconSize}
					style={{ marginLeft: iconMarginLeft, color: iconColor }}
					{...iconProps}
				/>
			</View>
		)
	}
	return <VectorIcon size={iconSize} style={[{ color: iconColor }, style]} {...iconProps} />
}
