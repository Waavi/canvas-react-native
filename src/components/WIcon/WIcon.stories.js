import React from 'react'
import { View, Text } from 'react-native'
import { storiesOf } from '@storybook/react-native'
import { select, number, color } from '@storybook/addon-knobs'
import { StoryView, StoryHeader } from '@modules/react-native-storybook/components'
import { Colors } from '@/theme'
import { WIcon } from './WIcon'

storiesOf('WIcon', module)
	.add(
		'all icons',
		() => (
			<StoryView scrollable>
				<StoryHeader.h1 text="Icon Catalog" />
				<View style={styles.container}>
					{icons.map(name => (
						<View key={name} style={styles.iconBox}>
							<View style={styles.iconWrapper}>
								<WIcon name={name} size={25} color={Colors.brand} />
							</View>
							<View style={styles.textWrapper}>
								<Text style={styles.text}>{name}</Text>
							</View>
						</View>
					))}
				</View>
			</StoryView>
		),
		{
			notes: 'Catalog of every icon',
		}
	)
	.add(
		'colors and sizes',
		() => {
			const name = select(
				'Icon',
				icons.reduce((prev, name) => ({ ...prev, [name]: name }), {}),
				'logo'
			)
			const size = number('Size', 100)
			const iconColor = color('Color', Colors.brand)
			return (
				<StoryView>
					<StoryHeader.h1
						text="Colors and sizes"
						description="Play with knobs to modify color and size"
					/>
					<WIcon name={name} size={size} color={iconColor} />
				</StoryView>
			)
		},
		{
			notes: 'An icon can be customized with color, size, width, height...',
		}
	)

const styles = {
	container: {
		flexDirection: 'row',
		flexWrap: 'wrap',
		justifyContent: 'center',
		alignItems: 'center',
	},
	iconBox: {
		borderWidth: 1,
		borderColor: '#aaa',
		width: 90,
		height: 90,
		padding: 5,
		justifyContent: 'center',
		alignItems: 'center',
	},
	iconWrapper: {
		flex: 2,
		justifyContent: 'center',
		alignItems: 'center',
	},
	textWrapper: {
		flex: 1,
	},
	text: {
		fontSize: 9,
		textAlign: 'center',
	},
}

const icons = [
	'logo',
	'logo-sm',
	'logo-o',
	'logo-o-sm',
	'logotype',
	'category-food',
	'category-shopping',
	'category-wellness',
	'category-other',
	'cards-search',
	'my-cards',
	'my-promos',
	'user',
	'user-o',
	'user-sm',
	'how-works',
	'qr',
	'share-circle',
	'legal',
	'logout',
	'nav-bars',
	'nav-actions',
	'search',
	'points',
	'offer',
	'stamps',
	'coupon',
	'bono',
	'heart',
	'heart-o',
	'heart-o-arrow',
	'tag',
	'glance-comments',
	'glance-plus',
	'glance-tick',
	'card',
	'photos',
	'card-catalog',
	'card-menu',
	'card-comments',
	'card-establishments',
	'promo-age-range',
	'promo-expiration',
	'exclamation-triangle',
	'promo-student',
	'thumb-up',
	'thumb-down',
	'oops',
	'tick-0',
	'tick-12',
	'tick-1',
	'tick-2',
	'tick-3',
	'filter-about-to-expire',
	'filter-last-visited',
	'filter-near-me',
	'start-transaction',
	'expand',
	'compress',
	'tick',
	'plus',
	'chevron-down',
	'chevron-up',
	'chevron-left',
	'chevron-right',
	'email',
	'warning-triangle',
	'lock',
	'no-cards',
	'no-promos',
	'no-comments',
	'exclamation',
	'times',
	'plus-light',
	'chevron-right-round',
	'chevron-left-round',
	'minus',
	'chevron-thin-right',
	'chevron-thin-up',
	'chevron-thin-left',
	'chevron-thin-down',
]
