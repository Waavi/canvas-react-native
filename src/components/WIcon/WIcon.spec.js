import React from 'react'
import { mount, shallow } from 'enzyme'
import { View } from 'react-native'
import { WIcon } from './WIcon'
import { VectorIcon } from './VectorIcon'

describe('Components/WIcon', () => {
	it('should use VectorIcon', () => {
		const props = { name: 'chevron' }
		const wrapper = shallow(<WIcon {...props} />)
		const innerText = wrapper.find(VectorIcon)
		expect(innerText.props()).toMatchObject(props)
	})
	it('should set a base style', () => {
		const wrapper = shallow(<WIcon />)
		const innerStyle = wrapper.find(VectorIcon).props().style
		expect(innerStyle).toBeTruthy()
		expect(innerStyle).toMatchSnapshot()
	})
	it('should have a "size" for each font family defined at Fonts catalog', () => {
		const sizes = [20, 22, 24]
		const wrapper = mount(
			<View>
				{sizes.map(size => (
					<WIcon key={size} size={sizes[size]} />
				))}
			</View>
		)
		expect(wrapper).toMatchSnapshot()
	})
	it('should have a "color" for each font family defined at Fonts catalog', () => {
		const colors = ['green', 'dark']
		const wrapper = mount(
			<View>
				{colors.map(color => (
					<WIcon key={color} color={colors[color]} />
				))}
			</View>
		)
		expect(wrapper).toMatchSnapshot()
	})
})
