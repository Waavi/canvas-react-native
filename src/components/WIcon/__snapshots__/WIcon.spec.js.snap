// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/WIcon should have a "color" for each font family defined at Fonts catalog 1`] = `
<View>
  <View>
    <WIcon
      key="green"
      size={20}
    >
      <Icon
        allowFontScaling={false}
        size={40}
        style={
          Array [
            Object {
              "color": "#222",
            },
            undefined,
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 40,
              },
              Array [
                Object {
                  "color": "#222",
                },
                undefined,
              ],
              Object {
                "fontFamily": "WelveIcons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 40,
                },
                Array [
                  Object {
                    "color": "#222",
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "WelveIcons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          />
        </Text>
      </Icon>
    </WIcon>
    <WIcon
      key="dark"
      size={20}
    >
      <Icon
        allowFontScaling={false}
        size={40}
        style={
          Array [
            Object {
              "color": "#222",
            },
            undefined,
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 40,
              },
              Array [
                Object {
                  "color": "#222",
                },
                undefined,
              ],
              Object {
                "fontFamily": "WelveIcons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 40,
                },
                Array [
                  Object {
                    "color": "#222",
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "WelveIcons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          />
        </Text>
      </Icon>
    </WIcon>
  </View>
</View>
`;

exports[`Components/WIcon should have a "size" for each font family defined at Fonts catalog 1`] = `
<View>
  <View>
    <WIcon
      key="20"
      size={20}
    >
      <Icon
        allowFontScaling={false}
        size={40}
        style={
          Array [
            Object {
              "color": "#222",
            },
            undefined,
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 40,
              },
              Array [
                Object {
                  "color": "#222",
                },
                undefined,
              ],
              Object {
                "fontFamily": "WelveIcons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 40,
                },
                Array [
                  Object {
                    "color": "#222",
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "WelveIcons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          />
        </Text>
      </Icon>
    </WIcon>
    <WIcon
      key="22"
      size={20}
    >
      <Icon
        allowFontScaling={false}
        size={40}
        style={
          Array [
            Object {
              "color": "#222",
            },
            undefined,
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 40,
              },
              Array [
                Object {
                  "color": "#222",
                },
                undefined,
              ],
              Object {
                "fontFamily": "WelveIcons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 40,
                },
                Array [
                  Object {
                    "color": "#222",
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "WelveIcons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          />
        </Text>
      </Icon>
    </WIcon>
    <WIcon
      key="24"
      size={20}
    >
      <Icon
        allowFontScaling={false}
        size={40}
        style={
          Array [
            Object {
              "color": "#222",
            },
            undefined,
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 40,
              },
              Array [
                Object {
                  "color": "#222",
                },
                undefined,
              ],
              Object {
                "fontFamily": "WelveIcons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 40,
                },
                Array [
                  Object {
                    "color": "#222",
                  },
                  undefined,
                ],
                Object {
                  "fontFamily": "WelveIcons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          />
        </Text>
      </Icon>
    </WIcon>
  </View>
</View>
`;

exports[`Components/WIcon should set a base style 1`] = `
Array [
  Object {
    "color": "#222",
  },
  undefined,
]
`;
